hail/src/main/scala/is/hail/expr/ir/streams/EmitStream.scala

```diff
@@ -1166,6 +1167,62 @@ object EmitStream {
           SStreamValue(producer)
         }
 
+      case StreamWhiten(stream, newChunkName, prevWindowName, vecSize, windowSize, chunkSize, blockSize, normalizeAfterWhiten) =>
+        produce(stream, cb).map(cb) { case blocks: SStreamValue =>
+          val state = new LocalWhitening(cb, SizeValueStatic(vecSize.toLong), windowSize.toLong, chunkSize.toLong, blockSize.toLong, outerRegion, normalizeAfterWhiten)
+          val eltType = blocks.st.elementType.asInstanceOf[SBaseStruct]
+          var resultField: SSettable = null
+
+          val blocksProducer = blocks.getProducer(cb.emb)
+          val producer: StreamProducer = new StreamProducer {
+            override def method: EmitMethodBuilder[_] = mb
+
+            override val length: Option[EmitCodeBuilder => Code[Int]] =
+              blocksProducer.length.map { l => (cb: EmitCodeBuilder) =>
+                val len = cb.memoize(l(cb))
+                len
+              }
+
+            override def initialize(cb: EmitCodeBuilder, outerRegion: Value[Region]): Unit = {
+              state.reset(cb)
+              blocksProducer.initialize(cb, outerRegion)
+            }
+
+            override val elementRegion: Settable[Region] = blocksProducer.elementRegion
+            override val requiresMemoryManagementPerElement: Boolean = blocksProducer.requiresMemoryManagementPerElement
+
+            override val LproduceElement: CodeLabel = mb.defineAndImplementLabel { cb =>
+              cb.goto(blocksProducer.LproduceElement)
+              cb.define(blocksProducer.LproduceElementDone)
+              val row = blocksProducer.element.toI(cb).get(cb, "StreamWhiten: missing tuple").asBaseStruct
+              row.loadField(cb, prevWindowName).consume(cb, {}, { prevWindow =>
+                state.initializeWindow(cb, prevWindow.asNDArray)
+              })
+              val block = row.loadField(cb, newChunkName).get(cb, "StreamWhiten: missing chunk").asNDArray
+              val whitenedBlock = LinalgCodeUtils.checkColMajorAndCopyIfNeeded(block, cb, elementRegion)
+              state.whitenBlock(cb, whitenedBlock)
+              // the 'newChunkName' field of 'row' is mutated in place and given
+              // to the consumer
+              val result = row.insert(cb, elementRegion, eltType.virtualType.asInstanceOf[TStruct], newChunkName -> EmitValue.present(whitenedBlock))
+              resultField = mb.newPField("StreamWhiten_result", result.st)
+              cb.assign(resultField, result)
```

Comment:
This feels pretty hacky, would welcome ideas on better ways to do this.

In the old version, this just mutated the incoming block in place. But for some reason, after merging with current main, the incoming block is row major (at least in my tests), forcing us to copy it. I was then having trouble getting the types right for the result after inserting the modified field, which is why I wait to define the result field until after I have the svalue to be stored.

Reply:
that seems fine. We do that elsewhere.